doc/workshop/12-external-idp-support.rst

```diff
@@ -0,0 +1,540 @@
+  May 06 10:43:28 client.ipademo.local kc.sh[27170]: 2022-05-06 10:43:28,412 INFO  [io.quarkus] (main) Profile prod activated.
+  May 06 10:43:28 client.ipademo.local kc.sh[27170]: 2022-05-06 10:43:28,412 INFO  [io.quarkus] (main) Installed features: [agroal…
+  Hint: Some lines were ellipsized, use -l to show in full.
+
+Now we can use it for setting up users and OAuth 2.0 clients. There are two
+handy scripts, ``kcadm.sh`` and ``kcreg.sh`` that allow to perform all
+operations without visiting Web UI.
+
+With ``kcadm.sh`` we login as admin and create user ``testuser1`` and set a password::
+
+  /opt/keycloak-18.0.0/bin/kcadm.sh config truststore \
+        --trustpass Secret123 \
+        /etc/pki/tls/private/keycloak.store
+
+  /opt/keycloak-18.0.0/bin/kcadm.sh config credentials \
+        --server https://$(hostname):8443/auth/ \
+        --realm master --user admin --password Secret123
+  Logging into https://client.ipademo.local:8443/auth/ as user admin of realm master
+
+  /opt/keycloak-18.0.0/bin/kcadm.sh create users \
+        -r master \
+        -s username=testuser1 -s enabled=true -s email=testuser1@ipademo.local
+  Created new user with id 'd319b32a-4cea-43c5-8ef8-19b2b8418d0a'
+
+  /opt/keycloak-18.0.0/bin/kcadm.sh set-password \
+        -r master \
+        --username testuser1 --new-password Secret123
+
+With ``kcreg.sh`` we can create OAuth 2.0 client using a pre-defined template
+that will include all parameters we need to allow OAuth 2.0 Device Authorization
+Grant flow::
+
+  /opt/keycloak-18.0.0/bin/kcreg.sh config credentials \
+        --server https://$(hostname):8443/auth \
+        --realm master --user admin --password Secret123
+
+  cat >ipa_client.json <<EOF
+  {
+    "enabled" : true,
+    "redirectUris" : [ "https://ipa-ca.$(hostname -d)/ipa/idp/*" ],
+    "webOrigins" : [ "https://ipa-ca.$(hostname -d)" ],
+    "protocol" : "openid-connect",
+    "publicClient" : true,
+    "attributes" : {
+      "oauth2.device.authorization.grant.enabled" : "true",
+      "oauth2.device.polling.interval": "5"
+    }
+  }
+  EOF
+
+  /opt/keycloak-18.0.0/bin/kcreg.sh create \
+        -f ipa_client.json  \
+        -s clientId=ipa_oidc_client
+
+At this point, we have a Keycloak instance with a default ``master`` realm
+(organization) and base URL ``https://$(hostname):8443/auth/``. In this realm we
+have created ``testuser1`` user with a simple password. We also created OAuth
+2.0 client ``ipa_oidc_client`` that is allowed to utilize OAuth 2.0 device
+authorization grant flow. This client has no client secret ("public OAuth 2.0
+client") associated. Confidential clients can also support device authorization
+grant flows.
+
+The client details include information about the redirect URIs. These are required
+to specify for public OAuth 2.0 clients, but they aren't used for OAuth 2.0
+device authorization grant flow.
+
+Two attributes specified in the OAuth 2.0 client definition for Keycloak enable
```

Comment:
Two attributes specified in the OAuth 2.0 client definition for Keycloak enable
OAuth 2.) device authorization grant processing and set the polling interval for
the client


Notice the two attributes specified in the OAuth 2.0 client definition for Keycloak: enable
oauth2.device.authorization.grant.enabled and set oauth2.device.polling.interval to 5 as the polling interval for
the client